src/mongocxx/insert_many_builder.cpp

```diff
@@ -0,0 +1,56 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <mongocxx/config/prelude.hpp>
+
+#include <mongocxx/insert_many_builder.hpp>
+
+#include <bsoncxx/builder/stream/document.hpp>
+
+#include <mongocxx/collection.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+insert_many_builder::insert_many_builder(const options::insert& options)
+    : _writes{options.ordered().value_or(true)}, _inserted_ids{}, _index{0} {
+    if (options.write_concern()) _writes.write_concern(*options.write_concern());
+};
+
+void insert_many_builder::operator()(const bsoncxx::document::view& doc) {
+    if (!doc["_id"]) {
+        bsoncxx::builder::stream::document new_document;
+        new_document << "_id" << bsoncxx::oid(bsoncxx::oid::init_tag)
+                     << bsoncxx::builder::stream::concatenate{doc};
+
+        _writes.append(model::insert_one{new_document.view()});
```

Comment:
FWIW I've been extremely wary of adding a requirement on a variant type because there is, as yet, no consensus at the standardization level about what semantics a variant type will have. It is definitely a troublesome aspect of working with views and values, but I don't want to make us dependent on MNMLSTC's particular choice of semantics for variants given its lack of forward progress and probably divergence from the standard.